snowfall/training/compute_expected_times.py

```diff
@@ -0,0 +1,183 @@
+#
+# Copyright (c)  2021  Xiaomi Corp.       (author: Fangjun Kuang)
+#
+
+import k2
+import torch
+
+
+def _create_phone_fsas(phone_seqs: k2.RaggedInt) -> k2.Fsa:
+    '''
+    Args:
+      phone_seqs:
+        It contains two axes with elements being phone IDs.
+        The last element of each sub-list is -1.
+    Returns:
+      Return an FsaVec representing the phone seqs.
+    '''
+    assert phone_seqs.num_axes() == 2
+    phone_seqs = k2.ragged.remove_values_eq(phone_seqs, -1)
+    return k2.linear_fsa(phone_seqs)
+
+
+def compute_expected_times_per_phone(mbr_lats: k2.Fsa,
+                                     ctc_topo: k2.Fsa,
+                                     dense_fsa_vec: k2.DenseFsaVec,
+                                     use_double_scores=True,
+                                     num_paths=100) -> torch.Tensor:
+    '''Compute expected times per phone in a n-best list.
+
+    See the following comments for more information:
+
+        - `<https://github.com/k2-fsa/snowfall/issues/96>`_
+        - `<https://github.com/k2-fsa/k2/issues/641>`_
+
+    Args:
+      mbr_lats:
+        An FsaVec.
+      ctc_topo:
+        The return value of :func:`build_ctc_topo`.
+      dense_fsa_vec:
+        It contains nnet_output.
+      use_double_scores:
+        True to use `double` in :func:`k2.random_paths`; false to use `float`.
+      num_paths:
+        Number of random paths to draw in :func:`k2.random_paths`.
+    Returns:
+      A 1-D torch.Tensor contains the expected times per pathphone_idx.
+    '''
+    lats = mbr_lats
+    assert len(lats.shape) == 3
+    assert hasattr(lats, 'phones')
+
+    # paths will be k2.RaggedInt with 3 axes: [seq][path][arc_pos],
+    # containing arc_idx012
+    paths = k2.random_paths(lats,
```

Comment:
I just found that `paths` is empty when `lats` is `mbr_lats` and device is cuda.

Will look into it.

Reply:
OK.  BTW did you already look at the decoding graph that you used to generate `den_lats`, and make sure that it has more nonzero labels than nonzero phones?

Reply:
Yes, I've added two print statements in this pull-request:
https://github.com/k2-fsa/snowfall/blob/f50e4f4b8d909d977c5373c884e293c46a3ddd8e/snowfall/training/mmi_mbr_graph.py#L162-L168

and the output is

```
den.phones torch.Size([30446]) nnz tensor(29928)
den.labels torch.Size([30446]) nnz tensor(30100)
```

`den.phones` contain more 0s than `den.labels`.